plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Stablecoin.hs

```diff
@@ -0,0 +1,385 @@
+    let fees = calcFees sc bs rate inpSCAction
+        (newState, newConstraints) = case inpSCAction of
+            MintStablecoin sc' ->
+                let scValue = stablecoinNominalPrice bs rate * (BC $ fromInteger $ unSC sc') in
+                (bs
+                { bsStablecoins = bsStablecoins bs + sc'
+                , bsReserves = bsReserves bs + fmap round (fees + scValue)
+                }, Constraints.mustForgeCurrency bsForgingPolicyScript scStablecoinTokenName (unSC sc'))
+            RedeemStablecoin sc' ->
+                let scValue = stablecoinNominalPrice bs rate * (BC $ fromInteger $ unSC sc') in
+                (bs
+                { bsStablecoins = bsStablecoins bs - sc'
+                , bsReserves = bsReserves bs + fmap round (fees - scValue)
+                }, Constraints.mustForgeCurrency bsForgingPolicyScript scStablecoinTokenName (unSC $ negate sc'))
+            MintReserveCoin rc ->
+                let rcValue = reservecoinNominalPrice sc bs rate * (BC $ fromInteger $ unRC rc) in
+                (bs
+                { bsReservecoins = bsReservecoins bs + rc
+                , bsReserves = bsReserves bs + fmap round (fees + rcValue)
+                }, Constraints.mustForgeCurrency bsForgingPolicyScript scReservecoinTokenName (unRC rc))
+            RedeemReserveCoin rc ->
+                let rcValue = reservecoinNominalPrice sc bs rate * (BC $ fromInteger $ unRC rc) in
+                (bs
+                { bsReservecoins = bsReservecoins bs - rc
+                , bsReserves = bsReserves bs + fmap round (fees - rcValue)
+                }, Constraints.mustForgeCurrency bsForgingPolicyScript scReservecoinTokenName (unRC $ negate rc))
+    guard $ isValidState sc newState rate
+    let dateConstraints = Constraints.mustValidateIn $ Interval.from obsSlot
+    pure (constraints <> newConstraints <> dateConstraints, newState)
+
+-- | A 'Value' with the given number of reservecoins
+reserveCoins :: Stablecoin -> Integer -> Value
+reserveCoins sc@Stablecoin{scReservecoinTokenName} =
+    let sym = Scripts.monetaryPolicyHash $ scriptInstance sc
+    in Value.singleton (Value.mpsSymbol sym) scReservecoinTokenName
+
+-- | A 'Value' with the given number of stablecoins
+stableCoins :: Stablecoin -> Integer -> Value
+stableCoins sc@Stablecoin{scStablecoinTokenName} =
+    let sym = Scripts.monetaryPolicyHash $ scriptInstance sc
+    in Value.singleton (Value.mpsSymbol sym) scStablecoinTokenName
+
+
+{-# INLINEABLE isValidState #-}
+isValidState :: Stablecoin -> BankState -> ConversionRate -> Bool
+isValidState sc bs@BankState{bsReservecoins, bsReserves, bsStablecoins} cr =
+    traceIfFalse "reservecoins < 0" (bsReservecoins >= RC 0)
+    && traceIfFalse "reserves < 0" (bsReserves >= BC 0)
+    && traceIfFalse "stablecoins < 0" (bsStablecoins >= SC 0)
+    && traceIfFalse "min reserves" (fmap fromInteger bsReserves >= minReserve sc cr bs)
```

Comment:
Is this actually impossible? Will this just mean you can't do anything with the machine if the price crashes and you go below the new min reserves?

Reply:
> Will this just mean you can't do anything with the machine if the price crashes and you go below the new min reserves?

Yes in that case you can't do anything except add more reserves. This is again from the Scala spec:

```scala
private def minReserve(sc: N): N = minReserveRatio * sc * oracle.conversionRate(peg, base)
```

I think it would be reasonable to treat the `minReserves` like the margin in the future contract: If the reserves drop below `minReserve` then stablecoin holders should still be able to redeem their coins but reservecoin holders would have to add more of the base currency before they could do anything

Reply:
> Is this actually impossible?

It's possible if the conversion rate is low enough. The state can go from valid to invalid simply by changing the conversion rate. This whole function is ported from Scala, maybe we don't need it at all - I *think* they should all follow from the transition function & there not being allowed any negative values on outputs

```scala
  def isValidTransaction(amountBase: N, amountSC: N, amountRC: N, feeInBase: N): Boolean = {
    val scValueInBase = amountSC * stablecoinNominalPrice(reserves, stablecoins)
    val rcValueInBase = amountRC * reservecoinNominalPrice(reserves, stablecoins, reservecoins)

    val newReserves = reserves - amountBase + feeInBase
    val newStablecoins = stablecoins + amountSC

    val correctPrices = { scValueInBase + rcValueInBase + amountBase == 0 }
    val correctFee = { feeInBase == (abs(amountBase) + abs(scValueInBase) + abs(rcValueInBase)) * fee }

    acceptableReserveChange(amountSC > 0, amountRC > 0, amountRC < 0, newReserves, newStablecoins) && correctPrices && correctFee
  }
```